File: samples/cea-techpulse/src/techNewsMcpService.ts
import { Client } from '@modelcontextprotocol/sdk/client/index.js';
import { StdioClientTransport } from '@modelcontextprotocol/sdk/client/stdio.js';

export interface TechNewsArticle {
  title: string;
  source: string;
  publishedAt: string;
  description: string;
  url: string;
}

export class TechNewsMcpService {
  private client: Client | null = null;
  private transport: StdioClientTransport | null = null;
  private isConnected = false;

  constructor(private mcpServerPath: string) {}

  async connect(): Promise<void> {
    if (this.isConnected) {
      return;
    }

    try {
      // Create transport that will spawn the MCP server process
      this.transport = new StdioClientTransport({
        command: 'node',
        args: [this.mcpServerPath],
        stderr: 'inherit',
        env: {
          ...process.env,
          // Add News API key if available from either environment variable name
          NEWS_API_KEY: process.env.NEWS_API_KEY || process.env.SECRET_NEWS_API_KEY || 'YOUR_API_KEY_HERE'
        }
      });

      this.client = new Client(
        {
          name: 'teams-ai-agent-news',
          version: '1.0.0',
        },
        {
          capabilities: {},
        }
      );

      // Connect to the server
      await this.client.connect(this.transport);
      this.isConnected = true;

      console.log('Connected to Tech News MCP Server');

    } catch (error) {
      console.error('Failed to connect to Tech News MCP server:', error);
      this.disconnect();
      throw error;
    }
  }

  async disconnect(): Promise<void> {
    if (this.client && this.transport) {
      await this.client.close();
    }

    this.client = null;
    this.transport = null;
    this.isConnected = false;
  }

  async getTechNews(category: 'general' | 'ai' | 'startups' | 'cybersecurity' | 'mobile' | 'gaming' = 'general', limit: number = 10): Promise<string> {
    if (!this.client || !this.isConnected) {
      throw new Error('Tech News MCP client not connected');
    }

    try {
      console.log(`Calling MCP server tool 'get_tech_news' with category: ${category}, limit: ${limit}`);
      const result = await this.client.callTool({
        name: 'get_tech_news',
        arguments: { category, limit }
      });

      console.log('MCP server response received for tech news');
      if (result.content && Array.isArray(result.content) && result.content.length > 0) {
        const firstContent = result.content[0] as any;
        const newsText = firstContent.text || 'No tech news available';
        console.log(`Tech news data length: ${newsText.length} characters`);
        return newsText;
      }

      return 'No tech news found';
    } catch (error) {
      console.error('Error getting tech news from MCP server:', error);
      throw new Error(`Failed to get tech news from MCP server: ${error}`);
    }
  }

  async searchTechNews(keyword: string, limit: number = 10, timeframe: 'today' | 'week' | 'month' = 'week'): Promise<string> {
    if (!this.client || !this.isConnected) {
      throw new Error('Tech News MCP client not connected');
    }

    try {
      console.log(`Calling MCP server tool 'search_tech_news' with keyword: ${keyword}, timeframe: ${timeframe}`);
      const result = await this.client.callTool({
        name: 'search_tech_news',
        arguments: { keyword, limit, timeframe }
      });

      console.log('MCP server response received for tech news search');
      if (result.content && Array.isArray(result.content) && result.content.length > 0) {
        const firstContent = result.content[0] as any;
        const searchResults = firstContent.text || 'No search results available';
        console.log(`Tech news search results length: ${searchResults.length} characters`);
        return searchResults;
      }

      return 'No search results found';
    } catch (error) {
      console.error('Error searching tech news from MCP server:', error);
      throw new Error(`Failed to search tech news from MCP server: ${error}`);
    }
  }

  async getTrendingTech(region: 'us' | 'gb' | 'ca' | 'au' = 'us', limit: number = 10): Promise<string> {
    if (!this.client || !this.isConnected) {
      throw new Error('Tech News MCP client not connected');
    }

    try {
      console.log(`Calling MCP server tool 'get_trending_tech' with region: ${region}, limit: ${limit}`);
      const result = await this.client.callTool({
        name: 'get_trending_tech',
        arguments: { region, limit }
      });

      console.log('MCP server response received for trending tech topics');
      if (result.content && Array.isArray(result.content) && result.content.length > 0) {
        const firstContent = result.content[0] as any;
        const trendingNews = firstContent.text || 'No trending topics available';
        console.log(`Trending tech topics data length: ${trendingNews.length} characters`);
        return trendingNews;
      }

      return 'No trending topics found';
    } catch (error) {
      console.error('Error getting trending tech topics from MCP server:', error);
      throw new Error(`Failed to get trending tech topics from MCP server: ${error}`);
    }
  }

  async getCompanyNews(company: string, limit: number = 8, timeframe: 'today' | 'week' | 'month' = 'week'): Promise<string> {
    if (!this.client || !this.isConnected) {
      throw new Error('Tech News MCP client not connected');
    }

    try {
      console.log(`Calling MCP server tool 'get_company_news' with company: ${company}, timeframe: ${timeframe}`);
      const result = await this.client.callTool({
        name: 'get_company_news',
        arguments: { company, limit, timeframe }
      });

      console.log('MCP server response received for company news');
      if (result.content && Array.isArray(result.content) && result.content.length > 0) {
        const firstContent = result.content[0] as any;
        const companyNews = firstContent.text || 'No company news available';
        console.log(`Company news data length: ${companyNews.length} characters`);
        return companyNews;
      }

      return 'No company news found';
    } catch (error) {
      console.error('Error getting company news from MCP server:', error);
      throw new Error(`Failed to get company news from MCP server: ${error}`);
    }
  }

  isServiceConnected(): boolean {
    return this.isConnected;
  }
}

File: samples/cea-techpulse-python/src/tech_news_mcp_service.py
from __future__ import annotations

import asyncio
import os
import sys
from mcp import ClientSession, StdioServerParameters
from mcp.client.stdio import stdio_client


class TechNewsMcpService:
    """Client wrapper around the Tech News MCP server (stdio transport)."""

    def __init__(self, mcp_server_path: str):
        self._mcp_server_path = mcp_server_path
        self._session: ClientSession | None = None
        self._client_cm = None  
        self._session_cm = None
        self._is_connected = False
        self._lock = asyncio.Lock()



    async def connect(self) -> None:
        async with self._lock:
            if self._is_connected:
                return

            env = {**os.environ, "NEWS_API_KEY": os.environ.get("NEWS_API_KEY", "")}

            server_params = StdioServerParameters(
                command=sys.executable,
                args=[self._mcp_server_path],
                env=env,
            )

            client_cm = stdio_client(server_params)
            try:
                read_stream, write_stream = await client_cm.__aenter__()
            except Exception:
                raise

            session_cm = ClientSession(read_stream, write_stream)
            try:
                session = await session_cm.__aenter__()
                await session.initialize()
            except Exception:
                await session_cm.__aexit__(None, None, None)
                await client_cm.__aexit__(None, None, None)
                raise

            self._client_cm = client_cm
            self._session_cm = session_cm
            self._session = session
            self._is_connected = True

    async def disconnect(self) -> None:
        async with self._lock:
            if self._session_cm:
                await self._session_cm.__aexit__(None, None, None)
            if self._client_cm:
                await self._client_cm.__aexit__(None, None, None)
            self._session = None
            self._client_cm = None
            self._session_cm = None
            self._is_connected = False

    def is_service_connected(self) -> bool:
        return self._is_connected



    async def _call_tool(self, name: str, arguments: dict) -> str:
        if not self._session or not self._is_connected:
            raise RuntimeError("Tech News MCP client not connected")

        result = await self._session.call_tool(name, arguments)

        if result.content:
            text = getattr(result.content[0], "text", None) or "No data available"
            return text
        return "No results found"



    async def get_tech_news(
        self,
        category: str = "general",
        limit: int = 10,
    ) -> str:
        return await self._call_tool("get_tech_news", {"category": category, "limit": limit})

    async def search_tech_news(
        self,
        keyword: str,
        limit: int = 10,
        timeframe: str = "week",
    ) -> str:
        return await self._call_tool(
            "search_tech_news", {"keyword": keyword, "limit": limit, "timeframe": timeframe}
        )

    async def get_trending_tech(
        self,
        region: str = "us",
        limit: int = 10,
    ) -> str:
        return await self._call_tool("get_trending_tech", {"region": region, "limit": limit})

    async def get_company_news(
        self,
        company: str,
        limit: int = 8,
        timeframe: str = "week",
    ) -> str:
        return await self._call_tool(
            "get_company_news", {"company": company, "limit": limit, "timeframe": timeframe}
        )
